prusti-viper/src/encoder/mir/specifications/interface.rs

```diff
@@ -44,36 +39,43 @@ pub(crate) trait SpecificationsInterface {
 
 impl<'v, 'tcx: 'v> SpecificationsInterface for super::super::super::Encoder<'v, 'tcx> {
     fn is_pure(&self, def_id: DefId) -> bool {
-        let mut cache = self.specifications_state.pure_cache.borrow_mut();
-        let result = cache.entry(def_id).or_insert_with(|| {
-            self.specifications_state
-                .specs
-                .borrow_mut()
-                .compute_is_pure(self.env(), def_id)
-        });
-
+        let result = self
+            .specifications_state
+            .specs
+            .borrow_mut()
+            .get_and_refine_proc_spec(self.env(), def_id)
+            .and_then(|spec| spec.pure.extract_refinements().inherit_refined())
+            .unwrap_or(false);
         trace!("is_pure {:?} = {}", def_id, result);
-        *result
+        result
     }
 
     fn is_trusted(&self, def_id: DefId) -> bool {
-        let mut cache = self.specifications_state.trusted_cache.borrow_mut();
-        let result = cache.entry(def_id).or_insert_with(|| {
-            self.specifications_state
-                .specs
-                .borrow_mut()
-                .compute_is_trusted(self.env(), def_id)
-        });
+        let result = self
+            .specifications_state
+            .specs
+            .borrow_mut()
+            .get_and_refine_proc_spec(self.env(), def_id)
+            .and_then(|spec| {
+                spec.trusted
+                    .extract_refinements()
+                    .with_selective_replacement()
```

Comment:
`is_pure` looks different that `is_trusted`: I'm guessing the intent is that `pure` is inherited but `trusted` is not. However, there is no way to declare a function as *not* trusted, correct? So if the base method is trusted, `with_selective_replacement` will always inherit the trustedness?

Reply:
Yes, I'm not aware of any way the user can mark a function as "not trusted".
The default, when there is no specification, is that a function is not trusted though (same for pure).

`with_selective_replacement` does not inherit anything, it ignores the possibly refined-from value. This is why there is the separate `inherit_refined` which can be used for the pure flag.

I am not sure whether we should handle `#[pure]` and `#[trusted]` equally in this case. When we attach an extern spec to a trait, the trait's specification will automatically be trusted, even if the user does not annotate it directly. The refinement would then inherit this trustedness on any local impl.

Reply:
Ok, thanks. I agree that `pure` and `trusted` should be treated differently. But I wrongly assumed `with_selective_replacement` would inherit. All is good.

Reply:
Hmm, the name maybe seems a bit "weird", but this is basically how "strategy (C) in the paper" is called (the code docs refer to that).

Apart from that, with the new structure of the `DefSpecificationMap`, we could also say that `ProcedureSpecification::{pure, trusted}` are `SpecificationItem::Empty` by default to account for non-user-typed specs and resolve it later.